sycl/plugins/level_zero/pi_level_zero.cpp

```diff
@@ -667,10 +707,20 @@ inline static pi_result createEventAndAssociateQueue(
 
   if (!ForceHostVisible)
     ForceHostVisible = DeviceEventsSetting == AllHostVisible;
-  PI_CALL(EventCreate(Queue->Context, Queue, ForceHostVisible, Event));
+
+  // If event is discarded then try to get event from the queue cache.
+  *Event = IsInternal ? Queue->getEventFromCache(ForceHostVisible) : nullptr;
+
+  if (*Event == nullptr)
+    PI_CALL(EventCreate(Queue->Context, Queue, ForceHostVisible, Event));
 
   (*Event)->Queue = Queue;
   (*Event)->CommandType = CommandType;
+  (*Event)->IsDiscarded = IsInternal;
+  // Discarded event doesn't own ze_event, it is used by multiple pi_event
+  // objects.
+  if (IsInternal)
+    (*Event)->OwnZeEvent = false;
```

Comment:
But we need to delete the L0 event at the end, how is this achieved?

Reply:
We achieve this by releasing all events from the cache at queue destruction. Event in the cache owns L0 event.
https://github.com/intel/llvm/pull/7256/files#diff-15dd1eb076d2164bd9e87d9057f05f652a716498e8cdf5975e564c65309a0985R3880-R3883

Reply:
OK, please tell it in the source comment here.

Reply:
Added a comment.